### PR TITLE
[aota] compiled forward outputs requires_grad alignment with eager

### PR DESCRIPTION
**Pull request by IvanKobzarev**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128016

Original issue: https://github.com/pytorch/pytorch/issues/114338

We assume only two possible mutually exclusive scenarios:

1. Running compiled region for training (Any of inputs has requires_grad)
	- Produced differentiable outputs should have requires_grad.

2. Running compiled region for inference (None of inputs has requires_grad)
	- All outputs do not have requires_grad.

Even if user runs the region under no_grad(), but has an input Tensor with requires_grad - we go Training scenario (1).

With current state that means:
1/ needs_autograd should not check torch.is_grad_enabled(), only that any of inputs requires_grad
2/ if needs_autograd => trace_joint (We are in training scenario 1.) => always run compiled region under with.enable_grad()

### CODE REVIEW

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 1]
```diff
@@ -199,7 +199,8 @@ def runtime_wrapper(args: List[Any]):
             for idx in indices_of_inps_to_detach:
                 if isinstance(args_[idx], torch.Tensor):
                     args_[idx] = args_[idx].detach()
-            with torch.autograd._force_original_view_tracking(True):
+            needs_grad = any(x.requires_grad for x in args_ if isinstance(x, Tensor))
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py:202]: I don't think we want to determine `need_grad` here by looking at if any of the inputs require grad. Instead, you can just assume that if `trace_joint == True`, we should be turning on `enable_grad` (because trace_joint implies that we need a training graph, and we need at least one output that requires grad.

Can you also add a comment explaining why we need the enable_grad at runtime? Namely that it's possible for a user to run a compiled region under no_grad, but still require some of the outputs of the graph to require gradients.

File: torch/_functorch/aot_autograd.py [hunk 2]
```diff
@@ -544,10 +544,7 @@ def convert(idx, x):
 
         fake_flat_args = process_inputs(flat_args)
 
-        needs_autograd = (
-            any(x.requires_grad for x in fake_flat_args if isinstance(x, Tensor))
-            and torch.is_grad_enabled()
-        )
+        needs_autograd = any(x.requires_grad for x in fake_flat_args if isinstance(x, Tensor))
```
Comment (bdhirsh) [inline torch/_functorch/aot_autograd.py:547]: This looks right! Tests please :) you can add them in `test_aotdispatch.py:TestAOTAutograd

Comment (bdhirsh) [reply torch/_functorch/aot_autograd.py:547]: also cc @jamesjwu - Ivan is working on another silent correctness AOTAutograd fix that boils down to "we traced an inference graph when we should have traced a training graph": https://github.com/pytorch/pytorch/issues/114338

**Comment by bdhirsh** [comment]
Also a PR description about the situation would be great :)

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 3]
```diff
@@ -199,7 +199,9 @@ def runtime_wrapper(args: List[Any]):
             for idx in indices_of_inps_to_detach:
                 if isinstance(args_[idx], torch.Tensor):
                     args_[idx] = args_[idx].detach()
-            with torch.autograd._force_original_view_tracking(True):
+            with torch.autograd._force_original_view_tracking(
+                True
+            ), torch.enable_grad():
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py:204]: Can you add a comment explaining why we need the `enable_grad` here (since it's pretty subtle - it's possible for a user to run a region of code under `no_grad()`, but still require us to compute gradients for it)

Comment (jamesjwu) [reply torch/_functorch/_aot_autograd/runtime_wrappers.py:206]: QQ: why doesn't the "enable_grad" in the subregion of the function apply `enable_grad`. I.e. why when calling compiled_fn, do we not automatically enable grad as needed? Why do we need to explicitly set it here? 

(This is assuming we still apply the other change in this PR, so that trace_joint is True when any elements require grad).

File: torch/_functorch/aot_autograd.py (hunk 2 above)
Comment (jamesjwu) [reply torch/_functorch/aot_autograd.py]: Do we need needs_autograd to be defined up here, instead of on line 573? i.e. why don't we refactor both of these if statements into a single check (perhaps in its own function)? 

Right now it's confusing that needs_autograd can be True on 547 and then set to false later, making it temporarily true

**Comment by bdhirsh** [comment]
> Even if user runs the region under no_grad(), but has an input Tensor with requires_grad - we go Training scenario (1).

Just to clarify: if an input tensor requires_grad but the entire region is executed under no_grad, that by itself does not mean that we should go down the training path (for example, running an entire model under no_grad() should normally go down the inference path).

The important bit here that caused the silent correctness is that there was an `enable_grad()` context manager nested underneath the no_grad, that caused some of our graph outputs to `require_grad=True` (even though `torch.is_grad_enabled()` was False at compile time!).

File: torch/_functorch/_aot_autograd/runtime_wrappers.py (hunk 3 above)
Comment (bdhirsh) [reply torch/_functorch/_aot_autograd/runtime_wrappers.py:206]: Yeah that is a good question. At runtime, we are effectively just executing a giant `torch.autograd.Function` containing our compiled fw and bw graphs (this is how we get training support).

But normally, the entire point of `no_grad` is to say "actually I do not want to compute any gradients". So without this change, if a user ran this code:
```
with torch.no_grad():
    return f(x)
```

That would effectively turn into this:
```
with torch.no_grad():
    return CompiledFunction.apply(x)
```

If `torch.is_grad_enabled() == False` when we run an `autograd.Function`, the `autograd.Function` will  respect the global grad state and all of its outputs will be set to not require grad. However, we know that we actually do need to compute some gradients here (even if the global grad state happens to be False).

Let me know if that is still unclear!

File: test/functorch/test_aotdispatch.py [hunk 4]
```diff
@@ -5291,6 +5291,25 @@ def f(a, b):
...
+            lambda: torch.ones(10, requires_grad=False),
+        ]
+
+        compiled_f = aot_function(fn, nop)
+        for inp_fn in inp_fns:
+            with torch.no_grad():
+                ref_out = fn(inp_fn())
+                out = compiled_f(inp_fn())
+
+                self.assertEqual(ref_out.requires_grad, out.requires_grad)
```
Comment (bdhirsh) [inline test/functorch/test_aotdispatch.py:5311]: nit: can you also test that our outputs + gradients are actually correct? A slightly more complicated test might be better:
```
def fn(x):
     out1 = x.sin()
     with torch.enable_grad():
         out2 = x.cos()
    return out1, out2

x = torch.randn(10, requires_grad=True)
with torch.no_grad():
    out1, out2 = f(x)
(out1 + out2).sum().backward()
# check that x.grad is the same compared to when we use `aot_function`
```

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 5]
```diff
@@ -199,7 +199,11 @@ def runtime_wrapper(args: List[Any]):
             for idx in indices_of_inps_to_detach:
                 if isinstance(args_[idx], torch.Tensor):
                     args_[idx] = args_[idx].detach()
-            with torch.autograd._force_original_view_tracking(True):
+            # We need enable_grad when user runs compiled region under no_grad,
+            # and inputs have requires_grad. Subregion of fn can be under enable_grad.
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py:203]: This comment doesn't quite explain what's going on:

(1) we know that if we `traced_joint == True`, we **must** be in the training case (and have some gradients to compute)

(2) It is possible to be in the `traced_joint == True` case even when the global grad state is disabled, via a user's `no_grad()` region (how? there can be a nested `enable_grad()` inside the function that forces some outputs to require gradients)


(3) therefore, whenever `trace_joint == True`, we need to unconditionally flip the grad state back on while we are executing the compiled function.

**Review by bdhirsh — APPROVED** [review]
pre-emptively stamping. Left a few more comments about better testing + code comment.

**Comment by eellison** [comment]
this regresses hf_Reformer from `2.4561 → 0.0172` in inference. do you want revert or forward fix ?
